akka-docs/src/main/paradox/stream/stream-refs.md

```diff
@@ -0,0 +1,156 @@
+
+Stream refs are not persistent, however it is simple to build a resumeable stream by introducing such protocol
+on the actor messaging layer. It pointing out that, that stream refs are absolutely expected to be sent over
+Akka remoting to other nodes within a cluster, and as such, complement and do not compete with plain Actor
+messaging. Actors would usually be used to establish the stream, by means of some initial message saying 
+"I want to offer you many log elements (the stream ref)", or alternatively in the opposite way "If you need
+to send me much data, here is the stream ref you can use to do so".   
+
+Since the two sides ("remote" and "local") of reach reference may be confusing to simply refer to by
+"remote" and "local", since either side can be seen as "local" or "remote" depending how we look at it,
+we propose to use the terminology "origin" and "target", which is defined by where the stream ref was created.
+For `SourceRef`s, the "origin" is the side which has the data that it is going to stream out. For `SinkRef`s
+the "origin" side is the actor system that is ready to receive the data and has allocated the ref. Those
+two may be seen as duals of each other, however to explain patterns about sharing references, we found this
+ wording to be rather useful.  
+
+### Source Refs - offering streaming data over network
+
+A @scala[@scaladoc[`SourceRef`](akka.stream.SourceRef)]@java[@javadoc[`SourceRef`](akka.stream.SourceRef)]
+can be offered to a remote actor system in order for it to consume some source of data that we have prepared
+locally. 
+
+Scala
+:   @@snip [FlowStreamRefsDocSpec.scala]($code$/scala/docs/stream/FlowStreamRefsDocSpec.scala) { #offer-source }
+
+The origin actor which creates and owns the Source could also perform some validation or additional setup
+when preparing the source. Once it has handed out the `SourceRef` the remote side can run it like this:
+
+Scala
+:   @@snip [FlowStreamRefsDocSpec.scala]($code$/scala/docs/stream/FlowStreamRefsDocSpec.scala) { #offer-source-use }
+
+
+@@@ warning
+  A `SourceRef` is *by design* "single-shot". i.e. it may only be materialized once.
+  This is in order to not complicate the mental model what materializing such value would mean.
+  
+  By being single-shot, we always know what it means, and on top of those semantics offer a fan-out
+  by emitting multiple `SourceRef`s which target the same `Source` that uses `Broadcast`.
+  This also allows for fine grained control how many streams a system can expect to be running 
+  at the same time, which is useful for capacity planning and "allowed number of concurrent streams
+  limiting" of clients.
+@@@
+
+
+![simple-graph-example.png](../images/source-ref-dance.png)
+
+### Sink Refs - offering to receive streaming data
+
+The dual of source references are A @scala[@scaladoc[`SourceRef`](akka.stream.SinkRef)]@java[@javadoc[`SourceRef`](akka.stream.SinkRef)]s. They can be used to offer the other side the capability to 
```

Comment:
scaladsl and javadsl in those class names? (might be more places)

Reply:
not applicable anymore :)